hudi-common/src/main/java/org/apache/hudi/common/model/HoodieAvroIndexedRecord.java

```diff
@@ -122,43 +152,54 @@ public Object[] getColumnValues(Schema recordSchema, String[] columns, boolean c
 
   @Override
   public Object getColumnValueAsJava(Schema recordSchema, String column, Properties props) {
+    setSchema(recordSchema);
     return AvroRecordContext.getFieldValueFromIndexedRecord(data, column);
   }
 
   @Override
   public HoodieRecord joinWith(HoodieRecord other, Schema targetSchema) {
+    setSchema(targetSchema);
     GenericRecord record = HoodieAvroUtils.stitchRecords((GenericRecord) data, (GenericRecord) other.getData(), targetSchema);
     return new HoodieAvroIndexedRecord(key, record, operation, metaData);
   }
 
   @Override
   public HoodieRecord prependMetaFields(Schema recordSchema, Schema targetSchema, MetadataValues metadataValues, Properties props) {
+    setSchema(recordSchema);
     GenericRecord newAvroRecord = HoodieAvroUtils.rewriteRecordWithNewSchema(data, targetSchema);
     updateMetadataValuesInternal(newAvroRecord, metadataValues);
     return new HoodieAvroIndexedRecord(key, newAvroRecord, operation, metaData);
   }
 
   @Override
   public HoodieRecord updateMetaField(Schema recordSchema, int ordinal, String value) {
+    setSchema(recordSchema);
     data.put(ordinal, value);
     return new HoodieAvroIndexedRecord(key, data, operation, metaData);
   }
 
   @Override
   public HoodieRecord rewriteRecordWithNewSchema(Schema recordSchema, Properties props, Schema newSchema, Map<String, String> renameCols) {
+    setSchema(recordSchema);
     GenericRecord record = HoodieAvroUtils.rewriteRecordWithNewSchema(data, newSchema, renameCols);
     return new HoodieAvroIndexedRecord(key, record, operation, metaData);
   }
 
   @Override
   public HoodieRecord truncateRecordKey(Schema recordSchema, Properties props, String keyFieldName) {
+    setSchema(recordSchema);
     ((GenericRecord) data).put(keyFieldName, StringUtils.EMPTY_STRING);
     return this;
   }
 
   @Override
   public boolean isDelete(Schema recordSchema, Properties props) {
-    return false;
+    if (getData().equals(SENTINEL)) {
+      return false; // Sentinel record is not a delete
+    }
+    setSchema(recordSchema);
+    DeleteContext deleteContext = new DeleteContext(props, recordSchema);
```

Comment:
Can we cache the `deleteContext`?

Reply:
I'm not sure where we would cache the context. Right now this will match our current performance with the payload records so I'm not sure it is worth the complexity of a cache.

Reply:
+1 its just a wrapper pojo - nothing heavy in constructor. caching is not needed.

Reply:
We already reuse the isDelete then should be fine.